py/emitinlinethumb.c

```diff
@@ -108,7 +108,7 @@ STATIC mp_uint_t emit_inline_thumb_count_params(emit_inline_asm_t *emit, mp_uint
             return 0;
         }
         const char *p = qstr_str(MP_PARSE_NODE_LEAF_ARG(pn_params[i]));
-        if (!(strlen(p) == 2 && p[0] == 'r' && p[1] == '0' + i)) {
+        if (!(strlen(p) == 2 && p[0] == 'r' && (mp_uint_t)p[1] == '0' + i)) {
```

Comment:
Just a question: C99 guarantees that `'0' + i` will be of type `mp_uint`?

Reply:
Yes, as far as I understand when you have signed + unsigned then the signed value is promoted to unsigned, and then they're widened to the widest integer, which in this case is mp_uint_t (which should be wider than char).

Reply:
Ugh, it's more complicated than that. First both sides are promoted to `int` if they're narrower or `unsigned int` if `int` can't represent all the values. Then if the widest operand is a signed type T and T is able to hold all values of the other type, the other type is promoted to T. Otherwise both sides are promoted to the unsigned type as wide as the widest of the two operands.

So here '0' is automatically converted to 'int', and then (assuming `mp_uint_t` is `unsigned int') both sides are promoted to 'unsigned int'. So the explanation is more complicated, but the result is the same ;)